okhttp-tests/src/test/java/okhttp3/internal/ws/RealWebSocketTest.java

```diff
@@ -83,418 +76,223 @@
     server.close(1000, "Bye!");
     assertTrue(client.processNextFrame());
     clientListener.assertTextMessage("Hello!");
     assertFalse(client.processNextFrame());
-    clientListener.assertClose(1000, "Bye!");
+    clientListener.assertClosing(1000, "Bye!");
   }
 
-  @Test public void protocolErrorBeforeCloseSendsClose() {
-    server2client.raw().write(ByteString.decodeHex("0a00")); // Invalid non-final ping frame.
+  @Test public void protocolErrorBeforeCloseSendsClose() throws IOException {
+    server2clientSink.write(ByteString.decodeHex("0a00")).emit(); // Invalid non-final ping frame.
 
     client.processNextFrame(); // Detects error, send close, close connection.
     assertTrue(clientConnectionClosed);
     clientListener.assertFailure(ProtocolException.class, "Control frames must be final.");
 
     server.processNextFrame();
-    serverListener.assertClose(1002, "");
+    serverListener.assertFailure(EOFException.class, null);
   }
 
   @Test public void protocolErrorInCloseResponseClosesConnection() throws IOException {
     client.close(1000, "Hello");
+    server.processNextFrame();
     assertFalse(clientConnectionClosed); // Not closed until close reply is received.
 
     // Manually write an invalid masked close frame.
-    server2client.raw().write(ByteString.decodeHex("888760b420bb635c68de0cd84f"));
+    server2clientSink.write(ByteString.decodeHex("888760b420bb635c68de0cd84f")).emit();
 
     client.processNextFrame(); // Detects error, closes connection immediately since close already sent.
     assertTrue(clientConnectionClosed);
     clientListener.assertFailure(ProtocolException.class, "Server-sent frames must not be masked.");
 
-    server.processNextFrame();
-    serverListener.assertClose(1000, "Hello");
-
+    serverListener.assertClosing(1000, "Hello");
     serverListener.assertExhausted(); // Client should not have sent second close.
   }
 
   @Test public void protocolErrorAfterCloseDoesNotSendClose() throws IOException {
     client.close(1000, "Hello!");
+    server.processNextFrame();
+
     assertFalse(clientConnectionClosed); // Not closed until close reply is received.
-    server2client.raw().write(ByteString.decodeHex("0a00")); // Invalid non-final ping frame.
+    server2clientSink.write(ByteString.decodeHex("0a00")).emit(); // Invalid non-final ping frame.
 
     client.processNextFrame(); // Detects error, closes connection immediately since close already sent.
     assertTrue(clientConnectionClosed);
     clientListener.assertFailure(ProtocolException.class, "Control frames must be final.");
 
-    server.processNextFrame();
-    serverListener.assertClose(1000, "Hello!");
+    serverListener.assertClosing(1000, "Hello!");
 
     serverListener.assertExhausted(); // Client should not have sent second close.
   }
 
-  @Test public void networkErrorReportedAsCloseNotFailure() {
-    server2client.close();
+  @Test public void networkErrorReportedAsCloseNotFailure() throws IOException {
```

Comment:
Indeed. https://github.com/square/okhttp/issues/2696

I think I disagree with that change. The WebSocket RFC doesn’t anticipate an `onFailure()` callback, and I think we should either keep `onFailure()` and use it universally, or go all-in on `onClose()` and synthesize codes 1005 and 1006 instead of supplying exceptions.